src/main/java/se/liu/ida/hefquin/engine/datastructures/impl/SolutionMappingsHashTable.java

```diff
@@ -1,85 +1,209 @@
 package se.liu.ida.hefquin.engine.datastructures.impl;
 
-import java.util.Iterator;
+import java.util.*;
 
 import org.apache.jena.graph.Node;
 import org.apache.jena.sparql.core.Var;
 
+import org.apache.jena.sparql.engine.binding.Binding;
 import se.liu.ida.hefquin.engine.data.SolutionMapping;
+import se.liu.ida.hefquin.engine.data.impl.SolutionMappingUtils;
 
 public class SolutionMappingsHashTable extends SolutionMappingsIndexBase
 {
+	// Having List<Node> as key type for the hash table is probably
+	// not the best choice in terms of efficiency. However, it will have
+	// to do for the moment.
+	// TODO: can this be made more efficient?
+	protected final Map<List<Node>, List<SolutionMapping>> map = new HashMap<>();
+	protected final Collection<Var> joinVariables;
+
+	public SolutionMappingsHashTable( final Set<Var> joinVariables ){
+		this.joinVariables = joinVariables;
+	}
+	public SolutionMappingsHashTable( final Var ... vars ) {
+		this.joinVariables = Arrays.asList(vars);
+	}
+
 	@Override
 	public int size() {
-		// TODO Auto-generated method stub
-		return 0;
+		int size = 0;
+		final Iterator<List<SolutionMapping>> li = map.values().iterator();
+		while( li.hasNext() ){
+			size = size + li.next().size();
+		}
+		return size;
 	}
 
 	@Override
 	public boolean isEmpty() {
-		// TODO Auto-generated method stub
-		return false;
+		return ! ( !map.isEmpty() && size()>0 );
```

Comment:
```suggestion
		for ( final List<SolutionMapping> l : map.values() ) {
			if ( ! l.isEmpty() ) {
				return false;
			}
		}
		return true;
```

Reply:
See my suggested change above. It is more efficient than your implementation. More precisely, for cases in which `map.isEmpty()` is `true`, they are equal. However, if `map.isEmpty()` is false, my suggested change does not need to iterate over all values of the `map`; instead, as soon as one value is found that is nonempty, the iteration can stop.